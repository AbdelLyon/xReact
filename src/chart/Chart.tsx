import {
  CategoryScale,
  ChartData,
  Chart as ChartJS,
  ChartOptions,
  ChartTypeRegistry,
  Legend,
  LinearScale,
  Title,
  Tooltip,
  BarController,
  DoughnutController,
  BarElement,
  ArcElement,
  PointElement,
  LineElement,
  RadialLinearScale,
  ScatterController,
  PolarAreaController,
  InteractionItem,
} from "chart.js";
import { useRef, useCallback } from "react";
import {
  Chart as ChartRoot,
  ChartProps,
  getElementAtEvent,
} from "react-chartjs-2";

// Register Chart.js components
ChartJS.register(
  CategoryScale,
  LinearScale,
  Title,
  Tooltip,
  Legend,
  RadialLinearScale,
  BarElement,
  ArcElement,
  PointElement,
  LineElement,
  BarController,
  DoughnutController,
  ScatterController,
  PolarAreaController,
);

type ChartType = keyof ChartTypeRegistry;

type Props<T extends ChartType> = {
  data: ChartData<T>;
  options?: ChartOptions<T>;
  type: T;
  getElementSelected?: (elementSelected: InteractionItem[]) => void;
} & ChartProps<T>;

export const Chart = <T extends ChartType>({
  type,
  data,
  options,
  getElementSelected,
  ...props
}: Props<T>) => {
  const chartRef = useRef(null);

  const handleClick = useCallback(
    (event: React.MouseEvent<HTMLCanvasElement>) => {
      if (chartRef.current) {
        const element = getElementAtEvent(chartRef.current, event);

        if (element.length && getElementSelected) {
          getElementSelected(element);
        }
      }
    },
    [getElementSelected],
  );

  return (
    <ChartRoot
      ref={chartRef}
      data={data}
      options={options}
      type={type}
      onClick={handleClick}
      {...props}
    />
  );
};
