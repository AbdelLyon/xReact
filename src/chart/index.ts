import "@/index.css";

export { Chart } from "./Chart";
export type { ChartOptions, ChartData, InteractionItem } from "chart.js";
