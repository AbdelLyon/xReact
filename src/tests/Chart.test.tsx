import { render, screen, fireEvent } from "@testing-library/react";
import { describe, it, expect, vi, beforeEach } from "vitest";
import {
  Chart as ChartJS,
  CategoryScale,
  LinearScale,
  Title,
  Tooltip,
  Legend,
  RadialLinearScale,
  BarElement,
  ArcElement,
  PointElement,
  LineElement,
  BarController,
  DoughnutController,
  ScatterController,
  PolarAreaController,
} from "chart.js";

import { Chart } from "@/chart";

// Register Chart.js components
ChartJS.register(
  CategoryScale,
  LinearScale,
  Title,
  Tooltip,
  Legend,
  RadialLinearScale,
  BarElement,
  ArcElement,
  PointElement,
  LineElement,
  BarController,
  DoughnutController,
  ScatterController,
  PolarAreaController,
);

describe("XuiChart", () => {
  const mockGetContext = vi
    .fn()
    .mockReturnValue({} as CanvasRenderingContext2D);

  HTMLCanvasElement.prototype.getContext = mockGetContext;

  beforeEach(() => {
    vi.mock("react-chartjs-2", async () => {
      const originalModule = await vi.importActual("react-chartjs-2");

      return {
        ...originalModule,
        getElementAtEvent: () => {
          return [{ index: 0, datasetIndex: 0 }];
        },
      };
    });
  });

  it("should match the snapshot", () => {
    const data = {
      labels: ["January", "February", "March"],
      datasets: [
        {
          label: "Dataset 1",
          data: [10, 20, 30],
        },
      ],
    };

    const { asFragment } = render(<Chart data={data} type="bar" />);

    expect(asFragment()).toMatchSnapshot();
  });

  it("should call getElementSelected when an element is clicked", () => {
    const data = {
      labels: ["January", "February", "March"],
      datasets: [
        {
          label: "Dataset 1",
          data: [10, 20, 30],
        },
      ],
    };

    const getElementSelected = vi.fn();

    render(
      <Chart data={data} getElementSelected={getElementSelected} type="bar" />,
    );

    const canvas = screen.getByRole("img");

    fireEvent.click(canvas);

    expect(getElementSelected).toHaveBeenCalledWith([
      { index: 0, datasetIndex: 0 },
    ]);
    expect(getElementSelected).toHaveBeenCalledTimes(1);
  });
});
