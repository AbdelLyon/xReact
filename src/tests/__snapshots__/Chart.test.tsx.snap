// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`XuiChart > should match the snapshot 1`] = `
<DocumentFragment>
  <canvas
    height="150"
    role="img"
    width="300"
  />
</DocumentFragment>
`;
